README
Introduction:

encoDHer is a python utility to facilitate symmetric encryption of email
messages using the Diffie-Hellman (DH) protocol for generating a shared
secret between the sender and recipient of an email message.

Keys are stored in the database file keys.db.

Installation requirements:

The encodher package requires python-gnupg-0.3.5 or greater to function.
The latest version can be obtained from:

http://pythonhosted.org/python-gnupg/

Additionally, the package makes use of the Diffie Hellman class
from Mark Loiseau:

http://blog.markloiseau.com/2013/01/diffie-hellman-tutorial-in-python/

Note that this Diffie Hellman class is included in the distribution
as dh.py.

Finally, this package requires the use of gnupg to sign DH public keys for
export and to identify DH public keys for import. You will need the
GPG secret key to sign your DH keys, and the receiver's GPG public key to
verify and import DH public keys.

Startup:

To use the program, you must first initialize the keys database.  This is
accomplished with the following command:

encodher.py --init

Once your database is created, you can generate keys for specific recipients
using the command:

encodher.py --gen-key sender@example.org receiver@otherexample.org

where your email address is sender@example.org and your intended recipient's
email address is receiver@otherexample.org. Note that using the DH shared
secret protocol requires generating a DH secret key for each sender -> receiver
pair.  So you will probably end up with several keys in the database.

Key distribution:

You can then export your signed DH public key with the command:

encodher.py --sign-pub sender@example.org receiver@otherexample.org

The output will be a signed version of your public key.  You can then
send that public key to a recipient over a non-secret channel.  An
anonymized key tablet will be developed later to facilitate DH public key
exchanges for the encodher utility, but any nonsecure channel will work.

When the recipient receives your DH public key, she can then import your
key into her database using the command:

encodher.py --import textfile

where textfile is a text file containing your signed DH public key. Note
that the GPG public key used to sign your DH public key must be in her
keychain in order to verify the signature on the DH public key.

Your intended recipient should then export her signed DH public key and
send it to you.  You can then import her signed DH public key and you
are ready to communicate using symmetric key encryption with a DH shared
secret key.

Message encryption and exchange:

To encrypt a message, first create the plaintext in a file.  You can
then encrypt the contents of that message with your DH shared secret
key using the command:

encodher.py --encode-email file sender@example.org receiver@otherexample.org

The email will be encrypted using symmetric aes256 encryption and output
to file.asc.  The file is an ascii-armored pgp-format file.

Upon receipt of the encrypted message, the receiver can then decrypt the
file using the command:

encodher.py --decode-email file.asc sender@example.org receiver@otherexample.org

The plaintext message will then be displayed.  It is probably good to note
here that the first email address is always the message sender, and the second
is always the receiver.  So for example, if I received a message at
alice@alice.com from my buddy bob@bob.com, the correct command to decrypt the
message would be:

encodher.py --decode-email file.asc bob@bob.com alice@alice.com

because bob was the sender of the email and alice was the receiver.  When I
encrypt an email to bob, the correct command is:

encodher.py --encode-email file alice@alice.com bob@bob.com

because in this case alice is the sender and bob the receiver.

Other options:

Various options for key management exist.  The following list of options
can be obtained by executing encodher.py without any arguments.

Options:
 --init, -i: initialize keys.db database
 --import, -m: import signed public key
 --mutate-key, -a: mutate DH secret key
 --sign-pub, -s: sign your own public key
 --change-toemail, -t: change toEmail on key
 --change-fromemail, -f: change fromEmail on key
 --change-pubkey, -p: change public key for fromEmail -> toEmail
 --encode-email, -e: symmetrically encode a file for fromEmail -> toEmail
 --decode-email, -d: symmetrically decode a file for fromEmail -> toEmail
 --list-keys, -l: list all keys in database
 --gen-secret, -c: generate shared secret for fromEmail -> toEmail
 --gen-key, -n: generate a new key for fromEmail -> toEmail
 --get-key, -g: get key for fromEmail -> toEmail from database
 --fetch-aam, -h: fetch messages from alt.anonymous.messages newsgroup

Perfect forward secrecy:

The primary reason for using symmetric encryption with DH shared secrets
for email exchanges is to provide for perfect forward secrecy (PFS).  If, after
an exchange, the DH keys are destroyed by both parties, any messages
encrypted with these keys can no longer be read, achieving PFS. Users can
change their keys to take advantage of PFS by executing:

encodher.py --mutate-key sender@example.org receiver@example.org

This will destroy sender@example.org's old DH secret key, generate a new
DH secret key and export the corresponding new DH public key encrypted with the
old DH shared secret.  This encrypted file can be emailed to the receiver,
who can then decrypt the file and import the key using the --import option.
If an unencrypted version of the new signed DH public key is desired,
the --sign-pub option can be used. Once the new DH public key has been
imported by the receiver, no one (including the original sender and receiver)
will be able to read the old messages. Change keys carefully.

Anonymous communication:

If anonymous communication is desired, encodher presents an option to encode
the email for transmission using the mixmaster network.  Once the appropriate
mixmaster headers are added (by answering affirmatively to the question about
sending anonymously), the message can be dispatched to the
alt.anonymous.messages newsgroup with the command:

mixmaster -c 1 < message.asc

where message.asc contains the encrypted message and plaintext headers. Note
that you must have the mixmaster package installed and stats updated for this
to work.

To receive messages sent to you at a.a.m using a DH key of yours, you can
execute the command:

encodher.py -h

This will fetch and decrypt all a.a.m messages for each of the keys in your
database.


CHANGELOG
0.3 - Oct. 16, 2013
 * keys.db is now encrypted with AES256 (passphrase supplied by user)
 * moved aam parser from rfc822 module to email module
 * --fetch-aam option now writes each encrypted message to a separate file
 * README.md and ANONYMOUS.md received significant updates
 * Updated encoDHer to use OpenSSL for key generation if M2Crypto is
   available. M2Crypto is a python wrapper library for OpenSSL. If M2Crypto
   is not available, encoDHer will use the old module from Mark Loiseau
   (renamed dh_pydhe and updated to use the same prime as in the distributed
   dhparams.pem file). A dhparams.pem file is now distributed with encoDHer
   that contains an 8192 bit prime using a generator of 5. Such a big prime
   is overkill for generating an AES256 encryption key, but what the heck...
   Unfortunately, old keys.db databases are not compatible with this upgrade
   and will need to be regenerated.
 * If the dhparams.pem file is absent, encoDHer will use the Loiseau module
   (renamed dh_legacy) with the original 6144-bit RFC3526 prime.  This is
   backwards-compatible with old keys.db databases, but will NOT be
   compatible with databases using the dh_m2crypto or dh_pydhe modules.

0.2 - Oct. 8, 2013
 * encoDHer now decrypts unicode strings properly
 * Invalid public key exception during a.a.m search is now handled
   gracefully (this might happen e.g. when you have created a new
   route, but not yet imported the destination public key)
 * s2k hash has been changed to sha256
 * --list-keys changed to --list-routes
 * added encodher DH and GPG public keys 

   $ gpg --fingerprint encodher
   pub   8192R/9F16F7B2 2013-10-09 [expires: 2018-10-08]
         Key fingerprint = BB8C 19BB 09C1 19FE 72AA  06F7 8BD1 334F 9F16 F7B2
   uid                  encodher
   sub   8192R/8E37D228 2013-10-09 [expires: 2018-10-08]
   sub   8192R/AD535BD8 2013-10-09 [expires: 2018-10-08]

0.1 - initial release


encodher.py
#!/usr/bin/env python
"""
encoDHer - a python package for symmetric encryption of email
using the Diffie-Hellman shared secret key protocol.

Copyright (C) 2013 by David R. Andersen

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

For more information, see https://github.com/rxcomm/encoDHer
"""
print 'encoDHer version 0.3'

import os
import sys
import hsub
import gnupg
import dhutils
import re
import time
import string
import random
import email
import nntplib
from getpass import getpass
from constants import *

gpg = gnupg.GPG(gnupghome=HOME, gpgbinary=GPGBINARY, keyring=KEYRING,
                secret_keyring=SECRET_KEYRING, options=['--throw-keyids',
                '--personal-digest-preferences=sha256','--s2k-digest-algo=sha256'])
gpg.encoding = 'utf-8'

try:
    opt = sys.argv[1]
except (IndexError):
    print 'Options:'
    print ' --init, -i: initialize keys.db database'
    print ' --import, -m: import signed public key'
    print ' --mutate-key, -a: mutate DH secret key for PFS'
    print ' --sign-pub, -s: sign your own public key'
    print ' --change-toemail, -t: change toEmail on key'
    print ' --change-fromemail, -f: change fromEmail on key'
    print ' --change-pubkey, -p: change public key for fromEmail -> toEmail'
    print ' --encode-email, -e: symmetrically encode a file for fromEmail -> toEmail'
    print ' --decode-email, -d: symmetrically decode a file for fromEmail -> toEmail'
    print ' --list-routes, -l: list all routes in database'
    print ' --gen-secret, -c: generate shared secret for fromEmail -> toEmail'
    print ' --gen-key, -n: generate a new key for fromEmail -> toEmail'
    print ' --get-key, -g: get key for fromEmail -> toEmail from database'
    print ' --fetch-aam, -h: fetch messages from alt.anonymous.messages newsgroup'
    print ' --clone-key, -y: clone key from one route to another'
    print ' --rollback, -b: roll back the a.a.m last read timestamp'
    print ' --change-dbkey, -k: change keys.db encryption key'
    sys.exit(0)

try:
    with open(KEYS_DB):
        dbpassphrase = getpass('Passphrase to decrypt keys.db: ')
except IOError:
    pass

def init():

    try:
        dhutils.initDB(gpg,dbpassphrase)
        print 'Database already exists.  Only the news timestamp was updated.'
    except NameError:
        dhutils.initDB(gpg,'password')
        print 'New keys.db database created, password is: password'
        print 'Change it immediately using: '+sys.argv[0]+' --change-dbkey'

def rollback():

    try:
        days = sys.argv[2]
    except (IndexError):
        print 'You need to supply the number of days to roll back nntp!'
        print 'Ex: '+sys.argv[0]+' --rollback <# of days>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    dhutils.rollback(days,gpg,dbpassphrase)

def importKey():

    try:
        file = sys.argv[2]
    except (IndexError):
        print 'You need to supply a source key file!'
        print 'Ex: '+sys.argv[0]+' --import <key file>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    print 'Importing new DH public key to database'

    with open (file, "r") as f:
        signed_data=f.read()

    verified = gpg.verify(str(signed_data))
    if verified.username is not None:
        print('Verified signed by: %s' % verified.username)
        print('at trust level: %s' % verified.trust_text)
    else:
        print 'Signature not valid'
        sys.exit(0)

    data = signed_data.split('\n')
    pubkey = ''
    for line in data:
        if len(line) == 50:
            pubkey += line
    while pubkey[:1] == '0':
        pubkey = pubkey[1:]

    try:
        toEmail = verified.username.split('<')[1].split('>')[0] # regular email
    except IndexError:
        toEmail = verified.username # only a name - probably anonymous
    print 'To Email is: %s' % toEmail.lower()
    fromEmail = raw_input('Enter From Email: ')

    keys = dhutils.getKeys(fromEmail.lower(),toEmail.lower(),gpg,dbpassphrase)

    if not keys:
        print 'key doesn\'t exist for the '+fromEmail+' -> '+toEmail+' route'
        print 'create new key?'
        ans = raw_input('y/N: ')
        if ans == 'y':
            dhutils.insertKeys(fromEmail.lower(),toEmail.lower(),pubkey.lower(),gpg,dbpassphrase)
    else:
        print 'key exists for the '+fromEmail.lower()+' -> '+toEmail.lower()+' route'
        print 'change key?'
        ans = raw_input('y/N: ')
        if ans == 'y':
            dhutils.changePubKey(fromEmail.lower(),toEmail.lower(),pubkey.lower(),gpg,dbpassphrase)


def sign_pub():

    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]

    except (IndexError):
        print 'You need to supply source and target email addresses!'
        print 'Ex: '+sys.argv[0]+' --sign-pub <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    privkey, mypubkey, otherpubkey = dhutils.getKeys(fromEmail,toEmail,gpg,dbpassphrase)
    while len(mypubkey) < 50*50:
        mypubkey = '0'+mypubkey
    brokenkey = [mypubkey[i:i+50] for i in range(0, len(mypubkey), 50)]
    new_mypubkey = ''
    for line in brokenkey:
        new_mypubkey += line+'\n'

    passphrase = getpass('Signing key ('+fromEmail+') password: ')
    signed_data = gpg.sign('DH Public Key:\n'+new_mypubkey+'\n', passphrase=passphrase,
                           keyid=fromEmail)
    print ''
    print str(signed_data)

    verified = gpg.verify(str(signed_data))
    if verified.username is not None:
        print('Verified signed by: %s' % verified.username)
        print('at trust level: %s' % verified.trust_text)
    else:
        print 'Sigature not verified'

def change_toEmail():
    try:
        fromEmail = sys.argv[2]
        oldToEmail = sys.argv[3]
        newToEmail = sys.argv[4]

    except (IndexError):
        print 'You need to supply fromEmail, oldToEmail, and newToEmail addresses'
        print 'Ex: '+sys.argv[0]+' --change-toemail <fromEmail> <oldToEmail> <newToEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    keys = dhutils.changeToEmail(fromEmail,oldToEmail,newToEmail,gpg,dbpassphrase)


def change_fromEmail():
    try:
        oldFromEmail = sys.argv[2]
        newFromEmail = sys.argv[3]
        toEmail = sys.argv[4]

    except (IndexError):
        print 'You need to supply oldFromEmail, newFromEmail, and toEmail addresses'
        print 'Ex: '+sys.argv[0]+' --change-fromemail <oldFromEmail> <newFromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    keys = dhutils.changeFromEmail(oldFromEmail,newFromEmail,toEmail,gpg,dbpassphrase)

def change_pub():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
        pubkey = sys.argv[4]

    except (IndexError):
        print 'You need to supply fromEmail, toEmail, and public key!'
        print 'Ex: '+sys.argv[0]+' --change-pub <fromEmail> <toEmail> <pub-key>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    keys = dhutils.changePubKey(fromEmail,toEmail,pubkey,gpg,dbpassphrase)

def hs():
    try:
        file_name = sys.argv[2]
        fromEmail = sys.argv[3]
        toEmail = sys.argv[4]
    except (IndexError):
        print 'You need to supply a target file to be encrypted, fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --encode-email <file> <fromEmail> <toEMail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    ans = raw_input('Do you want to send this message anonymously? (y/N)')
    if ans == 'y':
        sendAnon = True
    else:
        sendAnon = False

    passphrase = dhutils.genSharedSecret(fromEmail,toEmail,gpg,dbpassphrase)

    with open(file_name, "rb") as f:
        msg = gpg.encrypt_file(f, recipients=None, symmetric=CIPHER,
              always_trust=True, passphrase=passphrase)
        if sendAnon:
            iv = hsub.cryptorandom()
            hsubject = hsub.hash(passphrase[:16]) # first 64 bits to calc hsub

            # A note here about using part of the passphrase as the hsub password.
            # We use 64 bits (16 ascii bytes hex encoded = 8 bytes binary entropy)
            # for the hsub passphrase.  Assuming those 64 bits are completely
            # compromised (unlikekly, as it would require a rainbow table with
            # 3.4 x 10^38 entries) that leaves us with 192 bits of aes key entropy.
            # Still plenty strong.

    with open(file_name+'.asc', "w") as f:

        if sendAnon:
            f.write('To: mail2news@dizum.com,mail2news@m2n.mixmin.net\n')
            f.write('Subject: %s\n' % hsubject)
            f.write('Newsgroups: alt.anonymous.messages\n')
            f.write('X-No-Archive: Yes\n')
            f.write('\n')
        f.write(re.sub('\nV.*$', '', str(msg), count=1, flags=re.MULTILINE))

    print 'The encrypted file is '+file_name+'.asc'
    print 'Passphrase: %s' % passphrase


def hsd():
    # Try getting the route info from the X-encoDHer-Route header.
    # If the header doesn't exist, pass and parse the normal way.
    try:
        file_name = sys.argv[2]
        with open(file_name, 'r') as f:
            header = f.readline().split(': ')
            if header[0] == 'X-encoDHer-Route':
                route = header[1].strip().split('->')
                passphrase = dhutils.genSharedSecret(route[1],route[0],gpg,dbpassphrase)
                msg = gpg.decrypt_file(f, passphrase=passphrase, always_trust=True)
                if not msg:
                    print 'Bad shared secret!'
                    sys.exit(1)
                print '\n'+unicode(msg)
                sys.exit(0)
            else:
                pass
    except IndexError:
        pass

    try:
        file_name = sys.argv[2]
        fromEmail = sys.argv[3]
        toEmail = sys.argv[4]
    except (IndexError):
        print 'You need to supply a target file to be decrypted, and optionally route information!'
        print 'Ex: '+sys.argv[0]+' --decode-email <file> [<fromEmail> <toEmail>]'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    base, ext = os.path.splitext(file_name)
    with open(file_name, "r") as f:
        passphrase = dhutils.genSharedSecret(toEmail,fromEmail,gpg,dbpassphrase)
        msg = gpg.decrypt_file(f, passphrase=passphrase, always_trust=True)
        if not msg:
            print 'Bad shared secret!'
            sys.exit(1)
        print '\n'+unicode(msg)


def list_keys():
    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    print 'Listing of all routes in database:'
    dhutils.listKeys(gpg,dbpassphrase)

def secret():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --gen-secret <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    sharedSecret = dhutils.genSharedSecret(fromEmail,toEmail,gpg,dbpassphrase)
    print 'Secret: ',sharedSecret

def gen():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --gen-key <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    dhutils.makeKeys()
    dhutils.insertKeys(fromEmail,toEmail,1,gpg,dbpassphrase)


def get():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --get-key <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    privkey, mypubkey, otherpubkey = dhutils.getKeys(fromEmail,toEmail,gpg,dbpassphrase)
    print fromEmail+' Public Key: ', mypubkey
    print toEmail+' Public Key: ', otherpubkey

def delete():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --delete-key <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    dhutils.deleteKey(fromEmail,toEmail,gpg,dbpassphrase)

def mutate():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --mutate-key <fromEmail> <toEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    oldpassphrase = dhutils.genSharedSecret(fromEmail,toEmail,gpg,dbpassphrase)
    dhutils.mutateKey(fromEmail,toEmail,gpg,dbpassphrase)

    privkey, mypubkey, otherpubkey = dhutils.getKeys(fromEmail,toEmail,gpg,dbpassphrase)
    while len(mypubkey) < 50*50:
        mypubkey = '0'+mypubkey
    brokenkey = [mypubkey[i:i+50] for i in range(0, len(mypubkey), 50)]
    new_mypubkey = ''
    for line in brokenkey:
        new_mypubkey += line+'\n'

    passphrase = getpass('Signing key ('+fromEmail+') password: ')
    signed_data = gpg.sign('DH Public Key:\n'+new_mypubkey+'\n', passphrase=passphrase,
                            keyid=fromEmail)
    print ''
    print str(signed_data)

    ans = raw_input('Do you want to send this key anonymously? (y/N)')
    if ans == 'y':
        sendAnon = True
    else:
        sendAnon = False

    msg = gpg.encrypt(str(signed_data), recipients=None, symmetric=CIPHER,
          always_trust=True, passphrase=oldpassphrase)
    if sendAnon:
        iv = hsub.cryptorandom()
        hsubject = hsub.hash(oldpassphrase)

    with open('mutatedkey.asc', "w") as f:

        if sendAnon:
            f.write('To: mail2news@dizum.com,mail2news@m2n.mixmin.net\n')
            f.write('Subject: %s\n' % hsubject)
            f.write('Newsgroups: alt.anonymous.messages\n')
            f.write('X-No-Archive: Yes\n')
            f.write('\n')
        f.write(re.sub('\nV.*$', '', str(msg), count=1, flags=re.MULTILINE))
        print 'New key encrypted with old DH shared secret is in "mutatedkey.asc"'
        print 'Get unencrypted, signed copy of new key with '+sys.argv[0]+' --sign-pub '+fromEmail+' '+toEmail

def aam():
    GROUP  = "alt.anonymous.messages"

    timeStamp, curTimeStamp = dhutils.getNewsTimestamp(gpg,dbpassphrase)
    YYMMDD = time.strftime('%y%m%d', time.gmtime(timeStamp))
    HHMMSS = time.strftime('%H%M%S', time.gmtime(timeStamp))

    passphrases = dhutils.getListOfKeys(gpg,dbpassphrase)

    # connect to server
    server = nntplib.NNTP(NEWSSERVER,NEWSPORT)

    server.newnews(GROUP, YYMMDD, HHMMSS, '.newnews')

    with open ('.newnews', "r") as f:
        ids=f.read().splitlines()

        for msg_id in ids:
            try:
                resp, number, message_id, text = server.article(msg_id)
            except (nntplib.error_temp, nntplib.error_perm):
                pass # no such message (maybe it was deleted?)
            text = string.join(text, "\n")

            message = email.message_from_string(text)
            match = False

            for passphrase in passphrases:
                for label, item in message.items():
                    if label == 'Subject':
                        match = hsub.check(passphrase[2][:16],item)
                        #if match, decrypt
                        if match:
                            print '\nMail for: '+passphrase[0]+' from '+passphrase[1]
                            msg = gpg.decrypt(message.as_string(), passphrase=passphrase[2],
                                              always_trust=True)
                            if not msg:
                                print 'Bad shared secret!'
                                sys.exit(1)
                            print '\n'+unicode(msg)
                            with open('message_'+message_id[1:6]+'.txt', "w") as f:
                                f.write('X-encoDHer-Route: '+passphrase[1]+'->'+passphrase[0]+'\n')
                                f.write(message.as_string()+'\n')
                                print 'encrypted message stored in message_'+message_id[1:6]+'.txt'

    dhutils.setNewsTimestamp(curTimeStamp,gpg,dbpassphrase)
    print 'End of messages.'

def clone():
    try:
        fromEmail = sys.argv[2]
        toEmail = sys.argv[3]
        newFromEmail = sys.argv[4]
        newToEmail = sys.argv[5]
    except (IndexError):
        print 'You need to supply a fromEmail, and toEmail!'
        print 'Ex: '+sys.argv[0]+' --clone-key <oldFromEmail> <oldToEmail> <newFromEmail> <newToEmail>'
        sys.exit(1)

    try:
        with open(KEYS_DB): pass
    except IOError:
        print 'No keys database (keys.db)'
        print 'initialize the database with '+sys.argv[0]+' --init'
        sys.exit(1)

    dhutils.cloneKey(fromEmail,toEmail,newFromEmail,newToEmail,gpg,dbpassphrase)

def changeDBKey():

    dhutils.changeDBKey(KEYS_DB,gpg,dbpassphrase)
    print 'Key changed'

def errhandler():
    print 'Invalid option, try again!'
    print 'Execute '+sys.argv[0]+' to get a list of options.'
    sys.exit(1)

options = { '--init'   : init,
  '-i'   : init,
  '--import' : importKey,
  '-m' : importKey,
  '--sign-pub' : sign_pub,
  '-s' : sign_pub,
  '--change-toemail' : change_toEmail,
  '-t' : change_toEmail,
  '--change-fromemail' : change_fromEmail,
  '-f' : change_fromEmail,
  '--change-pubkey' : change_pub,
  '-p' : change_pub,
  '--encode-email' : hs,
  '-e' : hs,
  '--decode-email' : hsd,
  '-d' : hsd,
  '--list-routes' : list_keys,
  '-l' : list_keys,
  '--gen-secret' : secret,
  '-c' : secret,
  '--gen-key' : gen,
  '-n' : gen,
  '--get-key' : get,
  '-g' : get,
  '--delete-key' : delete,
  '-x' : delete,
  '--mutate-key' : mutate,
  '-a' : mutate,
  '--fetch-aam' : aam,
  '-h' : aam,
  '--clone-key' : clone,
  '-y' : clone,
  '--rollback' : rollback,
  '-b' : rollback,
  '--change-dbkey' : changeDBKey,
  '-k' : changeDBKey
}

def main():
    options.get(sys.argv[1],errhandler)()

if __name__ == '__main__':
    main()
